File: reimplementation/adl2/crates/adl-analysis/src/encode.rs
//! Region/bin encoding for the analysis engine.
//!
//! Reuses `adl-formula`'s region encoder at **statement granularity**:
//! each membership statement is encoded through a synthetic
//! single-statement region, so unsat cores name individual cuts and map
//! back to source lines (SPEC_ANALYSIS §3). The conjunction of the
//! per-statement formulas is the region formula by construction
//! (region encoding ≡ conjunction of statement encodings).
//!
//! Also hosts the documented **opaque-external re-tag pass**: for the
//! verifier, a comparison over an undeclared external function is an
//! exact atom over an opaque interned quantity (the per-event value
//! exists; identity is exact by interning), per the SPEC_ANALYSIS §2
//! model caveat ("opaque external-function values … are free"). The
//! interpreter keeps refusing them — one quantity, two consumers, two
//! honest answers.

use adl_formula::{DiagTable, EncodedRegion, Formula, Over, Under, encode_region};
use adl_sema::{Fragment, HKind, HNode, Hir, HirRegion, HirRegionStmt, Quantity, QuantityId};
use adl_solver::AssertName;
use adl_syntax::ast::CmpOp;
use adl_syntax::span::{LineMap, Span};
use std::collections::BTreeSet;

/// One encoded membership statement.
#[derive(Debug, Clone)]
pub struct StmtEnc {
    pub name: AssertName,
    pub span: Span,
    pub line: u32,
    pub text: String,
    pub formula: Formula,
    pub diags: DiagTable,
}

impl StmtEnc {
    #[must_use]
    pub fn over(&self) -> Over {
        self.formula.over()
    }

    #[must_use]
    pub fn under(&self) -> Under {
        self.formula.under()
    }
}

/// One region, encoded at statement granularity.
#[derive(Debug, Clone)]
pub struct RegionEnc {
    pub idx: usize,
    pub name: String,
    pub stmts: Vec<StmtEnc>,
    pub quantities: BTreeSet<QuantityId>,
    pub leaves_total: usize,
    pub leaves_encoded: usize,
    pub or_clauses: usize,
    pub dual_hedges: usize,
    /// (line, reason) for every Unknown leaf.
    pub dropped: Vec<(u32, String)>,
}

impl RegionEnc {
    #[must_use]
    pub fn exact(&self) -> bool {
        self.stmts.iter().all(|s| s.formula.is_exact())
    }
}

/// One bin set: a boundary-list `bin` statement (bins `[b0,b1), …,
/// [bn,∞)`) or the region's boolean-bin statements taken together.
#[derive(Debug, Clone)]
pub struct BinSetEnc {
    pub region_idx: usize,
    pub variable: String,
    pub bins: Vec<Formula>,
}

/// The encoded analysis unit.
#[derive(Debug, Clone)]
pub struct UnitEnc {
    pub regions: Vec<RegionEnc>,
    pub bin_sets: Vec<BinSetEnc>,
}

/// Verifier-side re-tag of undeclared-external-function quantities (see
/// module docs). Only `HKind::Quantity(ExternalFn)` nodes whose ONLY
/// problem is "not declared in the external library" are touched.
pub fn retag_opaque_externals(hir: &mut Hir) {
    let mut regions = std::mem::take(&mut hir.regions);
    for region in &mut regions {
        for stmt in &mut region.stmts {
            match stmt {
                HirRegionStmt::Select(n) | HirRegionStmt::Reject(n) | HirRegionStmt::Trigger(n) => {
                    retag_node(hir, n);
                }
                HirRegionStmt::Bin { var, .. } => retag_node(hir, var),
                HirRegionStmt::BinCond { cond, .. } => retag_node(hir, cond),
                _ => {}
            }
        }
    }
    hir.regions = regions;
}

fn retag_node(hir: &Hir, node: &mut HNode) {
    if let Fragment::Unsupported(reason) = &node.tag
        && reason.contains("is not declared in the external library")
        && let HKind::Quantity(q) = &node.kind
        && matches!(hir.table.quantity(*q), Quantity::ExternalFn { .. })
    {
        node.tag = Fragment::InFragment;
    }
    match &mut node.kind {
        HKind::Neg(a) | HKind::Not(a) | HKind::Abs(a) => retag_node(hir, a),
        HKind::Binary { lhs, rhs, .. } | HKind::Cmp { lhs, rhs, .. } => {
            retag_node(hir, lhs);
            retag_node(hir, rhs);
        }
        HKind::And(v) | HKind::Or(v) => {
            for n in v {
                retag_node(hir, n);
            }
        }
        HKind::Band { expr, .. } => retag_node(hir, expr),
        HKind::Ternary { guard, then, els } => {
            retag_node(hir, guard);
            retag_node(hir, then);
            if let Some(e) = els {
                retag_node(hir, e);
            }
        }
        _ => {}
    }
}

/// Encode a synthetic single-purpose region and remove it again. Only
/// the quantity table keeps the (harmless) interned growth.
fn encode_synthetic(hir: &mut Hir, stmts: Vec<HirRegionStmt>, span: Span) -> EncodedRegion {
    let name = hir.symbols.intern("__adl2_synth__");
    hir.regions.push(HirRegion { name, stmts, span });
    hir.region_name_order.push(name);
    let idx = hir.regions.len() - 1;
    let enc = encode_region(hir, idx);
    hir.regions.pop();
    hir.region_name_order.pop();
    enc
}

fn stmt_span(stmt: &HirRegionStmt) -> Span {
    match stmt {
        HirRegionStmt::Select(n) | HirRegionStmt::Reject(n) | HirRegionStmt::Trigger(n) => n.span,
        HirRegionStmt::Inherit { span, .. }
        | HirRegionStmt::Bin { span, .. }
        | HirRegionStmt::BinCond { span, .. }
        | HirRegionStmt::NonMembership { span, .. } => *span,
    }
}

fn is_membership(stmt: &HirRegionStmt) -> bool {
    match stmt {
        HirRegionStmt::Select(_)
        | HirRegionStmt::Reject(_)
        | HirRegionStmt::Inherit { .. }
        | HirRegionStmt::Trigger(_) => true,
        HirRegionStmt::Bin { .. } | HirRegionStmt::BinCond { .. } => false,
        // Unsupported non-membership statements still contribute an
        // Unknown leaf (honest coverage); in-fragment ones contribute
        // nothing.
        HirRegionStmt::NonMembership { tag, .. } => !tag.is_in_fragment(),
    }
}

/// Walk a formula counting coverage; `Dual` counts once (a hedge, not a
/// drop) and is not descended into.
fn coverage(f: &Formula, diags: &DiagTable, out: &mut RegionEnc, map: &LineMap) {
    match f {
        Formula::True | Formula::False | Formula::Atom(_) => {
            out.leaves_total += 1;
            out.leaves_encoded += 1;
        }
        Formula::And(v) => {
            for p in v {
                coverage(p, diags, out, map);
            }
        }
        Formula::Or(v) => {
            out.or_clauses += 1;
            for p in v {
                coverage(p, diags, out, map);
            }
        }
        Formula::Unknown(d) => {
            out.leaves_total += 1;
            if let Some(diag) = diags.get(*d) {
                let (line, _) = map.line_col(diag.span.start);
                out.dropped.push((line, diag.reason.clone()));
            }
        }
        Formula::Dual { .. } => {
            out.leaves_total += 1;
            out.leaves_encoded += 1;
            out.dual_hedges += 1;
        }
    }
}

/// Every quantity in a formula, both Dual branches included.
pub fn formula_quantities(f: &Formula, out: &mut BTreeSet<QuantityId>) {
    match f {
        Formula::True | Formula::False | Formula::Unknown(_) => {}
        Formula::Atom(a) => out.extend(a.terms().iter().map(|&(_, q)| q)),
        Formula::And(v) | Formula::Or(v) => {
            for p in v {
                formula_quantities(p, out);
            }
        }
        Formula::Dual { plus, minus, .. } => {
            formula_quantities(plus, out);
            formula_quantities(minus, out);
        }
    }
}

/// Encode every region (statement granularity) and every bin set.
pub fn encode_unit(hir: &mut Hir, src: &str) -> UnitEnc {
    let map = LineMap::new(src);
    let mut regions = Vec::new();
    let mut bin_sets = Vec::new();

    for ridx in 0..hir.regions.len() {
        let name = hir.symbols.display(hir.regions[ridx].name).to_owned();
        let stmt_list: Vec<HirRegionStmt> = hir.regions[ridx].stmts.clone();
        let mut enc = RegionEnc {
            idx: ridx,
            name,
            stmts: Vec::new(),
            quantities: BTreeSet::new(),
            leaves_total: 0,
            leaves_encoded: 0,
            or_clauses: 0,
            dual_hedges: 0,
            dropped: Vec::new(),
        };

        let mut cond_bins: Vec<Formula> = Vec::new();
        for (sidx, stmt) in stmt_list.iter().enumerate() {
            let span = stmt_span(stmt);
            if is_membership(stmt) {
                let e = encode_synthetic(hir, vec![stmt.clone()], span);
                let (line, _) = map.line_col(span.start);
                // Source-line text, or — for a merged cross-file unit with no
                // single `src` — the canonical HIR render of the cut, so the
                // unsat-core explanation stays meaningful instead of blank.
                let text = if src.is_empty() {
                    match stmt {
                        HirRegionStmt::Select(n) | HirRegionStmt::Trigger(n) => {
                            adl_sema::render_node(hir, n)
                        }
                        HirRegionStmt::Reject(n) => {
                            format!("reject {}", adl_sema::render_node(hir, n))
                        }
                        _ => String::new(),
                    }
                } else {
                    map.line_text(src, span.start).trim().to_owned()
                };
                coverage(&e.formula, &e.diags, &mut enc, &map);
                formula_quantities(&e.formula, &mut enc.quantities);
                enc.stmts.push(StmtEnc {
                    name: AssertName::new(format!("R{ridx}S{sidx}")),
                    span,
                    line,
                    text,
                    formula: e.formula,
                    diags: e.diags,
                });
            } else {
                match stmt {
                    HirRegionStmt::Bin { var, edges, .. } => {
                        if let Some(set) = encode_boundary_bins(hir, ridx, var, edges, span, src) {
                            bin_sets.push(set);
                        }
                    }
                    HirRegionStmt::BinCond { cond, .. } => {
                        let e =
                            encode_synthetic(hir, vec![HirRegionStmt::Select(cond.clone())], span);
                        cond_bins.push(e.formula);
                    }
                    _ => {}
                }
            }
        }
        if !cond_bins.is_empty() {
            bin_sets.push(BinSetEnc {
                region_idx: ridx,
                variable: "boolean bins".to_owned(),
                bins: cond_bins,
            });
        }
        regions.push(enc);
    }

    UnitEnc { regions, bin_sets }
}

/// `bin v b0 … bn` ⇒ formulas for `[b0,b1), …, [bn-1,bn), [bn,∞)`
/// (real-valued edges, open last bin — SPEC_LANGUAGE §4.3, divergence 5).
fn encode_boundary_bins(
    hir: &mut Hir,
    region_idx: usize,
    var: &HNode,
    edges: &[String],
    span: Span,
    src: &str,
) -> Option<BinSetEnc> {
    if edges.is_empty() {
        return None;
    }
    // Bin variable label from source text, or — for a merged unit with no
    // single `src` — the canonical HIR render (instead of a useless `?`).
    let variable = src
        .get(var.span.start as usize..var.span.end as usize)
        .map(|s| s.trim().to_owned())
        .unwrap_or_else(|| adl_sema::render_node(hir, var));
    let num = |text: &str| HNode::new(HKind::Num(text.to_owned()), span);
    let cmp = |op: CmpOp, rhs: HNode| {
        HNode::new(
            HKind::Cmp {
                op,
                lhs: Box::new(var.clone()),
                rhs: Box::new(rhs),
            },
            span,
        )
    };
    let mut bins = Vec::new();
    for i in 0..edges.len() {
        let mut stmts = vec![HirRegionStmt::Select(cmp(CmpOp::Ge, num(&edges[i])))];
        if i + 1 < edges.len() {
            stmts.push(HirRegionStmt::Select(cmp(CmpOp::Lt, num(&edges[i + 1]))));
        }
        let e = encode_synthetic(hir, stmts, span);
        bins.push(e.formula);
    }
    Some(BinSetEnc {
        region_idx,
        variable,
        bins,
    })
}

File: cpp/libs/analysis/src/encode.cpp
#include "adl2/analysis/encode.hpp"

#include "adl2/sema/dump.hpp"
#include "adl2/sema/ops.hpp"

#include <algorithm>
#include <cctype>
#include <memory>
#include <utility>

namespace adl2::analysis {
namespace {

using adl2::formula::DiagTable;
using adl2::formula::EncodedRegion;
using adl2::formula::Formula;
using adl2::sema::CmpOp;
using adl2::sema::Fragment;
using adl2::sema::HNode;
using adl2::sema::Hir;
using adl2::sema::HirRegion;
using adl2::sema::HirRegionStmt;
using adl2::sema::Quantity;
using adl2::sema::QuantityId;
using adl2::sema::QuantityKind;
using adl2::sema::QuantityTable;
using adl2::sema::Span;

/// Byte-offset → line map. Local copy of syntax LineMap so analysis never
/// includes parser headers.
struct LineMap {
  std::vector<std::size_t> line_starts;

  explicit LineMap(const std::string& src) {
    line_starts.push_back(0);
    for (std::size_t i = 0; i < src.size(); ++i) {
      if (src[i] == '\n') line_starts.push_back(i + 1);
    }
  }

  std::pair<std::uint32_t, std::uint32_t> line_col(std::size_t offset) const {
    auto it = std::upper_bound(line_starts.begin(), line_starts.end(), offset);
    std::size_t line = static_cast<std::size_t>(it - line_starts.begin());
    if (line == 0) line = 1;
    --line;
    std::uint32_t col = static_cast<std::uint32_t>(offset - line_starts[line] + 1);
    return {static_cast<std::uint32_t>(line + 1), col};
  }

  std::string line_text(const std::string& src, std::size_t offset) const {
    auto lc = line_col(offset);
    std::size_t start = line_starts[lc.first - 1];
    std::size_t end = (lc.first < line_starts.size()) ? line_starts[lc.first] : src.size();
    std::string s = src.substr(start, end - start);
    while (!s.empty() && (s.back() == '\n' || s.back() == '\r')) s.pop_back();
    return s;
  }
};

std::string trim(std::string s) {
  auto not_space = [](unsigned char c) { return !std::isspace(c); };
  s.erase(s.begin(), std::find_if(s.begin(), s.end(), not_space));
  s.erase(std::find_if(s.rbegin(), s.rend(), not_space).base(), s.end());
  return s;
}

void retag_node(const Hir& hir, HNode& node) {
  if (!node.tag.is_in_fragment() &&
      node.tag.reason.find("is not declared in the external library") != std::string::npos &&
      node.kind == HNode::Kind::Quantity &&
      hir.table.quantity(node.qid).kind == QuantityKind::ExternalFn) {
    node.tag = Fragment::ok();
  }
  switch (node.kind) {
    case HNode::Kind::Neg:
    case HNode::Kind::Not:
    case HNode::Kind::Abs:
      if (node.a) retag_node(hir, *node.a);
      break;
    case HNode::Kind::Binary:
    case HNode::Kind::Cmp:
      if (node.a) retag_node(hir, *node.a);
      if (node.b) retag_node(hir, *node.b);
      break;
    case HNode::Kind::And:
    case HNode::Kind::Or:
      for (auto& n : node.items) retag_node(hir, n);
      break;
    case HNode::Kind::Band:
      if (node.a) retag_node(hir, *node.a);
      break;
    case HNode::Kind::Ternary:
      if (node.a) retag_node(hir, *node.a);
      if (node.b) retag_node(hir, *node.b);
      if (node.c) retag_node(hir, *node.c);
      break;
    default:
      break;
  }
}

EncodedRegion encode_synthetic(Hir& hir, std::vector<HirRegionStmt> stmts, Span span) {
  HirRegion syn;
  syn.name = hir.symbols.intern("__adl2_synth__");
  syn.stmts = std::move(stmts);
  syn.span = span;
  hir.regions.push_back(std::move(syn));
  hir.region_name_order.push_back(hir.regions.back().name);
  std::size_t idx = hir.regions.size() - 1;
  EncodedRegion enc = adl2::formula::encode_region(hir, idx);
  hir.regions.pop_back();
  hir.region_name_order.pop_back();
  return enc;
}

HirRegionStmt opaque_stmt(Span span, const char* reason) {
  HirRegionStmt s;
  s.kind = HirRegionStmt::Kind::NonMembership;
  s.nm_kind = "inherit";
  s.tag = Fragment::unsupported(reason);
  s.span = span;
  return s;
}

void flatten_inherits(const Hir& hir, const std::vector<HirRegionStmt>& stmts,
                      std::vector<std::size_t>& stack, std::vector<HirRegionStmt>& out) {
  for (const auto& stmt : stmts) {
    if (stmt.kind != HirRegionStmt::Kind::Inherit) {
      out.push_back(stmt);
      continue;
    }
    if (stmt.region >= hir.regions.size()) {
      out.push_back(opaque_stmt(stmt.span, "reference to an unknown region"));
      continue;
    }
    bool cycle = false;
    for (std::size_t s : stack) {
      if (s == stmt.region) {
        cycle = true;
        break;
      }
    }
    if (cycle) {
      out.push_back(opaque_stmt(stmt.span, "region inheritance cycle"));
      continue;
    }
    std::vector<HirRegionStmt> inherited;
    for (const auto& s : hir.regions[stmt.region].stmts) {
      if (s.kind == HirRegionStmt::Kind::Bin || s.kind == HirRegionStmt::Kind::BinCond) {
        continue;
      }
      inherited.push_back(s);
    }
    stack.push_back(stmt.region);
    flatten_inherits(hir, inherited, stack, out);
    stack.pop_back();
  }
}

Span stmt_span(const HirRegionStmt& stmt) {
  switch (stmt.kind) {
    case HirRegionStmt::Kind::Select:
    case HirRegionStmt::Kind::Reject:
    case HirRegionStmt::Kind::Trigger:
      return stmt.node.span;
    case HirRegionStmt::Kind::Inherit:
    case HirRegionStmt::Kind::Bin:
    case HirRegionStmt::Kind::BinCond:
    case HirRegionStmt::Kind::NonMembership:
      return stmt.span;
  }
  return stmt.span;
}

bool is_membership(const HirRegionStmt& stmt) {
  switch (stmt.kind) {
    case HirRegionStmt::Kind::Select:
    case HirRegionStmt::Kind::Reject:
    case HirRegionStmt::Kind::Inherit:
    case HirRegionStmt::Kind::Trigger:
      return true;
    case HirRegionStmt::Kind::Bin:
    case HirRegionStmt::Kind::BinCond:
      return false;
    case HirRegionStmt::Kind::NonMembership:
      return !stmt.tag.is_in_fragment();
  }
  return false;
}

bool is_presence_atom(const QuantityTable& table, const Formula& f) {
  if (f.kind != Formula::Kind::Atom) return false;
  if (f.atom.terms().size() != 1) return false;
  return table.quantity(f.atom.terms()[0].second).kind == QuantityKind::Present;
}

void coverage(const QuantityTable& table, const Formula& f, const DiagTable& diags,
              RegionEnc& out, const LineMap& map) {
  switch (f.kind) {
    case Formula::Kind::True:
    case Formula::Kind::False:
    case Formula::Kind::Atom:
      if (is_presence_atom(table, f)) return;
      out.leaves_total += 1;
      out.leaves_encoded += 1;
      return;
    case Formula::Kind::And:
      for (const auto& p : f.items) coverage(table, p, diags, out, map);
      return;
    case Formula::Kind::Or:
      out.or_clauses += 1;
      for (const auto& p : f.items) coverage(table, p, diags, out, map);
      return;
    case Formula::Kind::Unknown: {
      out.leaves_total += 1;
      if (const auto* diag = diags.get(f.diag)) {
        auto lc = map.line_col(diag->span.start);
        out.dropped.push_back({lc.first, diag->reason});
      }
      return;
    }
    case Formula::Kind::Dual:
      out.leaves_total += 1;
      out.leaves_encoded += 1;
      out.dual_hedges += 1;
      return;
  }
}

std::optional<BinSetEnc> encode_boundary_bins(Hir& hir, std::size_t region_idx,
                                              const HNode& var,
                                              const std::vector<std::string>& edges,
                                              Span span, const std::string& src) {
  if (edges.empty()) return std::nullopt;
  std::string variable;
  if (var.span.start < src.size() && var.span.end <= src.size() &&
      var.span.end >= var.span.start) {
    variable = trim(src.substr(var.span.start, var.span.end - var.span.start));
  }
  if (variable.empty()) variable = adl2::sema::render_node(hir, var);

  auto num = [&](const std::string& text) {
    HNode n = HNode::make(HNode::Kind::Num, span);
    n.text = text;
    return n;
  };
  auto cmp = [&](CmpOp op, HNode rhs) {
    HNode n = HNode::make(HNode::Kind::Cmp, span);
    n.cmp = op;
    n.a = std::make_unique<HNode>(var);
    n.b = std::make_unique<HNode>(std::move(rhs));
    return n;
  };

  BinSetEnc set;
  set.region_idx = region_idx;
  set.variable = std::move(variable);
  for (std::size_t i = 0; i < edges.size(); ++i) {
    std::vector<HirRegionStmt> stmts;
    HirRegionStmt ge;
    ge.kind = HirRegionStmt::Kind::Select;
    ge.node = cmp(CmpOp::Ge, num(edges[i]));
    ge.span = span;
    stmts.push_back(std::move(ge));
    if (i + 1 < edges.size()) {
      HirRegionStmt lt;
      lt.kind = HirRegionStmt::Kind::Select;
      lt.node = cmp(CmpOp::Lt, num(edges[i + 1]));
      lt.span = span;
      stmts.push_back(std::move(lt));
    }
    EncodedRegion e = encode_synthetic(hir, std::move(stmts), span);
    set.bins.push_back(std::move(e.formula));
  }
  return set;
}

}  // namespace

void retag_opaque_externals(Hir& hir) {
  for (auto& region : hir.regions) {
    for (auto& stmt : region.stmts) {
      switch (stmt.kind) {
        case HirRegionStmt::Kind::Select:
        case HirRegionStmt::Kind::Reject:
        case HirRegionStmt::Kind::Trigger:
        case HirRegionStmt::Kind::Bin:
        case HirRegionStmt::Kind::BinCond:
          retag_node(hir, stmt.node);
          break;
        default:
          break;
      }
    }
  }
}

void formula_quantities(const Formula& f, std::set<QuantityId>& out) {
  switch (f.kind) {
    case Formula::Kind::True:
    case Formula::Kind::False:
    case Formula::Kind::Unknown:
      return;
    case Formula::Kind::Atom:
      for (const auto& t : f.atom.terms()) out.insert(t.second);
      return;
    case Formula::Kind::And:
    case Formula::Kind::Or:
      for (const auto& p : f.items) formula_quantities(p, out);
      return;
    case Formula::Kind::Dual:
      if (f.plus) formula_quantities(*f.plus, out);
      if (f.minus) formula_quantities(*f.minus, out);
      return;
  }
}

UnitEnc encode_unit(Hir& hir, const std::string& src) {
  LineMap map(src);
  UnitEnc unit;

  for (std::size_t ridx = 0; ridx < hir.regions.size(); ++ridx) {
    std::string name = hir.symbols.display(hir.regions[ridx].name);
    std::vector<HirRegionStmt> own = hir.regions[ridx].stmts;
    std::vector<HirRegionStmt> stmt_list;
    stmt_list.reserve(own.size());
    std::vector<std::size_t> stack{ridx};
    flatten_inherits(hir, own, stack, stmt_list);

    RegionEnc enc;
    enc.idx = ridx;
    enc.name = std::move(name);

    std::vector<Formula> cond_bins;
    for (std::size_t sidx = 0; sidx < stmt_list.size(); ++sidx) {
      const HirRegionStmt& stmt = stmt_list[sidx];
      Span span = stmt_span(stmt);
      if (is_membership(stmt)) {
        EncodedRegion e = encode_synthetic(hir, {stmt}, span);
        auto lc = map.line_col(span.start);
        std::uint32_t line = span.line != 0 ? span.line : lc.first;
        std::string text;
        if (src.empty()) {
          switch (stmt.kind) {
            case HirRegionStmt::Kind::Select:
            case HirRegionStmt::Kind::Trigger:
              text = adl2::sema::render_node(hir, stmt.node);
              break;
            case HirRegionStmt::Kind::Reject:
              text = "reject " + adl2::sema::render_node(hir, stmt.node);
              break;
            default:
              break;
          }
        } else {
          text = trim(map.line_text(src, span.start));
        }
        coverage(hir.table, e.formula, e.diags, enc, map);
        formula_quantities(e.formula, enc.quantities);
        StmtEnc se;
        se.name = adl2::solver::AssertName::make("R" + std::to_string(ridx) + "S" +
                                                 std::to_string(sidx));
        se.span = span;
        se.line = line;
        se.text = std::move(text);
        se.formula = std::move(e.formula);
        se.diags = std::move(e.diags);
        enc.stmts.push_back(std::move(se));
      } else if (stmt.kind == HirRegionStmt::Kind::Bin) {
        if (auto set = encode_boundary_bins(hir, ridx, stmt.node, stmt.edges, span, src)) {
          unit.bin_sets.push_back(std::move(*set));
        }
      } else if (stmt.kind == HirRegionStmt::Kind::BinCond) {
        HirRegionStmt sel;
        sel.kind = HirRegionStmt::Kind::Select;
        sel.node = stmt.node;
        sel.span = span;
        EncodedRegion e = encode_synthetic(hir, {std::move(sel)}, span);
        cond_bins.push_back(std::move(e.formula));
      }
    }
    if (!cond_bins.empty()) {
      BinSetEnc set;
      set.region_idx = ridx;
      set.variable = "boolean bins";
      set.bins = std::move(cond_bins);
      unit.bin_sets.push_back(std::move(set));
    }
    unit.regions.push_back(std::move(enc));
  }
  return unit;
}

}  // namespace adl2::analysis
